### PR TITLE
Add new test for blocking static field access during class init

### PR DESCRIPTION
**Pull request by gacholio**
The existing test isn't rigorous enough - static field accesses were
invalidly allowed during certain class initialization patterns.

See: #2794

[ci skip]

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

### CODE REVIEW

File: test/functional/VM_Test/src/j9vm/test/clinit/GetStaticTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	static {
+		GetStaticTest.write("bad");
+		try { Thread.sleep(5000); } catch(Throwable t) { throw new RuntimeException(t); }
+		GetStaticTest.write("good");
+	}
+}
+
+public class GetStaticTest {
+	public static boolean passed = false;
+    public static void write(String a) { GetStaticTestHelper.i = a; }
```
Comment (DanHeidinga) [inline test/functional/VM_Test/src/j9vm/test/clinit/GetStaticTest.java]: tabs vs spaces

File: test/functional/VM_Test/src/j9vm/test/clinit/GetStaticTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+package j9vm.test.clinit;
+
+class GetStaticTestHelper {
+	public static String i;
+
+	static {
+		GetStaticTest.write("bad");
+		try { Thread.sleep(5000); } catch(Throwable t) { throw new RuntimeException(t); }
```
Comment (DanHeidinga) [inline test/functional/VM_Test/src/j9vm/test/clinit/GetStaticTest.java]: Can you break that into multiple lines?

File: test/functional/VM_Test/src/j9vm/test/clinit/GetStaticTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+			public void run() {
+				try { Thread.sleep(2000); } catch(Throwable t) { throw new RuntimeException(t); }
+				String value = read();
+				passed = value.equals("good");
+			}
+		};
+		t.start();
+		// If running count=0, force read() to be compiled and GetStaticTestHelper to be initialized
+		read();
+		try { t.join(); } catch (Throwable e) { throw new RuntimeException(e); }
```
Comment (DanHeidinga) [inline test/functional/VM_Test/src/j9vm/test/clinit/GetStaticTest.java]: multiple lines please - much easier to debug

File: test/functional/VM_Test/src/j9vm/test/clinit/GetStaticTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+		if (!passed) {
+			throw new RuntimeException("TEST FAILED");
+		}
+	}
+
+	public static void main(String[] args) { GetStaticTest.test(); }
```
Comment (DanHeidinga) [inline test/functional/VM_Test/src/j9vm/test/clinit/GetStaticTest.java]: Split to multiple lines

**Comment by gacholio** [comment]
I'll think some more about using locks vs sleep.

**Comment by gacholio** [comment]
jenkins test sanity xlinux jdk8

**Comment by DanHeidinga** [comment]
An alternative is to force the deadlock and detect it with a MXBean like: https://docs.oracle.com/javase/6/docs/api/java/lang/management/ThreadMXBean.html#findDeadlockedThreads%28%29

Thoughts?

**Comment by gacholio** [comment]
I doubt the mxbean would detect this kind of deadlock (it's a wait/notify style deadlock, not a enter deadlock) - seems like too much engineering for this simple case.

**Comment by gacholio** [comment]
Closing this in favour of fixing the existing tests with (mostly) locks.

**Pull request closed without merging**